massa-versioning-worker/src/versioning_middleware.rs

```diff
@@ -0,0 +1,156 @@
+use std::collections::{HashMap, VecDeque};
+
+use massa_models::amount::Amount;
+use massa_time::{MassaTime, TimeError};
+
+use crate::versioning::{Advance, MipStore};
+use massa_versioning_exports::VersioningError;
+
+/// Struct used to keep track of announced versions in previous blocks
+pub struct VersioningMiddleware {
+    latest_announcements: VecDeque<u32>,
+    counts: HashMap<u32, usize>,
+    nb_blocks_considered: usize,
+    mip_store: MipStore,
+}
+
+impl VersioningMiddleware {
+    /// Creates a new empty versioning middleware.
+    pub fn new(nb_blocks_considered: usize, mip_store: MipStore) -> Self {
+        VersioningMiddleware {
+            latest_announcements: VecDeque::with_capacity(nb_blocks_considered + 1),
+            counts: HashMap::new(),
+            nb_blocks_considered,
+            mip_store,
+        }
+    }
+
+    /// Add the announced version of a new block
+    /// If needed, remove the info related to a ancient block
+    pub fn new_block(&mut self, version: u32) {
```

Comment:
I would suggest to use a lib for that like:  https://docs.rs/queues/latest/queues/.

This could make the code more readable...

Reply:
Indead, a [queues](https://docs.rs/queues/latest/queues/index.html)::[CircularBuffer](https://docs.rs/queues/latest/queues/struct.CircularBuffer.html) is probably a good thing to use in this case for readability.